File: backend/app/api/audits.py
"""Audit endpoints: create, poll, stats, history, delete, report.

Privacy model:
  * Anonymous audits (user_id is NULL) are shareable by UUID — this powers
    the "no signup required" flow and copied share links.
  * User-owned audits are private: only the owning user may view, report,
    or delete them. Accessing another user's audit returns 403.
"""
from __future__ import annotations

import uuid
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_current_user_optional
from app.core.database import get_db
from app.core.rate_limit import audit_per_hour, audit_per_minute, client_key
from app.models import Audit, User
from app.schemas import AuditCreate, AuditListItem, AuditOut
from app.workers.queue import enqueue_audit

router = APIRouter(prefix="/audits", tags=["audits"])


def _serialize_audit(audit: Audit, previous_score: Optional[int] = None) -> dict:
    data = {
        "public_id": audit.public_id,
        "url": audit.url,
        "status": audit.status,
        "progress": audit.progress,
        "stage": audit.stage,
        "max_pages": audit.max_pages,
        "language": audit.language,
        "overall_score": audit.overall_score,
        "category_scores": audit.category_scores,
        "summary": audit.summary,
        "results": audit.results,
        "ai_recommendations": audit.ai_recommendations,
        "error_message": audit.error_message,
        "error_code": audit.error_code,
        "started_at": audit.started_at,
        "completed_at": audit.completed_at,
        "previous_score": previous_score,
        "score_change": (audit.overall_score - previous_score) if (audit.overall_score is not None and previous_score is not None) else None,
    }
    return data


def _get_previous_score(db: Session, audit: Audit) -> Optional[int]:
    if audit.website_id is None:
        return None
    prev = (
        db.query(Audit)
        .filter(
            Audit.website_id == audit.website_id,
            Audit.id != audit.id,
            Audit.status == "completed",
            Audit.overall_score.isnot(None),
        )
        .order_by(Audit.completed_at.desc())
        .first()
    )
    return prev.overall_score if prev else None


def _get_audit_for_access(db: Session, public_id: str, user: Optional[User]) -> Audit:
    """Return an audit the current user may access, or raise 404/403."""
    audit = db.query(Audit).filter(Audit.public_id == public_id).first()
    if not audit:
        raise HTTPException(status_code=404, detail="Audit not found.")
    # User-owned audits are private. Anonymous audits remain shareable by UUID.
    if audit.user_id is not None and (user is None or user.id != audit.user_id):
        raise HTTPException(status_code=403, detail="You do not have access to this audit.")
    return audit


@router.post("", response_model=AuditOut, status_code=status.HTTP_201_CREATED)
def create_audit(
    payload: AuditCreate,
    request: Request,
    db: Session = Depends(get_db),
    user: Optional[User] = Depends(get_current_user_optional),
):
    key = client_key(request)
    if not audit_per_minute.allow(key):
        raise HTTPException(status_code=429, detail="You are starting audits too quickly. Please wait a moment.")
    if not audit_per_hour.allow(key):
        raise HTTPException(status_code=429, detail="You have reached the hourly audit limit. Please try again later.")

    # Server-side monthly plan limit (database is the source of truth).
    max_pages = None
    if user is not None:
        from app.services.billing import get_monthly_usage, get_or_create_subscription, plan_limits

        sub = get_or_create_subscription(db, user)
        limits = plan_limits(sub.plan)
        used = get_monthly_usage(db, user)
        if used >= limits["monthly_audits"]:
            raise HTTPException(
                status_code=403,
                detail={
                    "code": "monthly_limit_reached",
                    "message": "You have reached your monthly audit limit. Upgrade to Pro to continue.",
                },
            )
        max_pages = limits["max_pages"]

    audit = Audit(
        public_id=str(uuid.uuid4()),
        user_id=user.id if user else None,
        url=payload.url.strip(),
        status="queued",
        progress=0,
        stage="Queued",
        max_pages=max_pages,
        language=payload.language,
    )
    db.add(audit)
    db.commit()
    db.refresh(audit)

    if not enqueue_audit(audit.id):
        audit.status = "failed"
        audit.error_message = "Our auditors are busy right now. Please try again in a moment."
        audit.error_code = "crawl_failed"
        db.commit()
        db.refresh(audit)

    return _serialize_audit(audit)


@router.get("", response_model=List[AuditListItem])
def list_audits(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    audits = (
        db.query(Audit)
        .filter(Audit.user_id == user.id)
        .order_by(Audit.started_at.desc())
        .limit(100)
        .all()
    )
    items = []
    for audit in audits:
        previous = _get_previous_score(db, audit)
        items.append(
            {
                "public_id": audit.public_id,
                "url": audit.url,
                "status": audit.status,
                "overall_score": audit.overall_score,
                "started_at": audit.started_at,
                "score_change": (audit.overall_score - previous) if (audit.overall_score is not None and previous is not None) else None,
                "partial": bool(audit.results and audit.results.get("partial")),
                "error_code": audit.error_code,
                "error_message": audit.error_message,
            }
        )
    return items


@router.get("/stats")
def audit_stats(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    """Real statistics for the authenticated user's dashboard."""
    audits = db.query(Audit).filter(Audit.user_id == user.id).all()
    completed = [a for a in audits if a.status == "completed" and a.overall_score is not None]
    scores = [a.overall_score for a in completed]
    recent = sorted(audits, key=lambda a: a.started_at, reverse=True)[:5]
    return {
        "total_audits": len(audits),
        "completed_audits": len(completed),
        "average_score": round(sum(scores) / len(scores)) if scores else None,
        "best_score": max(scores) if scores else None,
        "recent_audits": [
            {
                "public_id": a.public_id,
                "url": a.url,
                "status": a.status,
                "overall_score": a.overall_score,
                "started_at": a.started_at,
                "partial": bool(a.results and a.results.get("partial")),
                "error_code": a.error_code,
            }
            for a in recent
        ],
    }


@router.get("/{public_id}", response_model=AuditOut)
def get_audit(
    public_id: str,
    db: Session = Depends(get_db),
    user: Optional[User] = Depends(get_current_user_optional),
):
    audit = _get_audit_for_access(db, public_id, user)
    return _serialize_audit(audit, _get_previous_score(db, audit))


@router.get("/{public_id}/report", response_model=AuditOut)
def get_report(
    public_id: str,
    db: Session = Depends(get_db),
    user: Optional[User] = Depends(get_current_user_optional),
):
    audit = _get_audit_for_access(db, public_id, user)
    if audit.status != "completed":
        raise HTTPException(status_code=409, detail="This audit has not finished yet.")
    return _serialize_audit(audit, _get_previous_score(db, audit))


@router.delete("/{public_id}")
def delete_audit(public_id: str, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    audit = _get_audit_for_access(db, public_id, user)
    if audit.user_id is None or audit.user_id != user.id:
        raise HTTPException(status_code=403, detail="You can only delete your own audits.")
    db.delete(audit)
    db.commit()
    return {"ok": True}

File: backend/src/routes/audits.ts
import { Router } from "express";
import { ObjectId } from "mongodb";
import { randomUUID } from "node:crypto";
import { config } from "../config.js";
import { collections, toId, type AuditDoc } from "../db.js";
import { asyncHandler, optionalAuth, requiredAuth, type AuthRequest } from "../middleware/auth.js";
import { auditPerHour, auditPerMinute, clientKey } from "../rateLimit.js";
import { enqueueAudit } from "../queue.js";
import { URLValidationError, validateUrl } from "../services/crawler/urlValidator.js";

export const auditsRouter = Router();

function serializeAudit(audit: AuditDoc, previousScore?: number | null) {
  const scoreChange =
    audit.overallScore != null && previousScore != null ? audit.overallScore - previousScore : null;
  return {
    public_id: audit.publicId,
    url: audit.url,
    status: audit.status,
    progress: audit.progress,
    stage: audit.stage,
    max_pages: audit.maxPages,
    language: audit.language,
    overall_score: audit.overallScore ?? null,
    category_scores: audit.categoryScores ?? null,
    summary: audit.summary ?? null,
    results: audit.results ?? null,
    ai_recommendations: audit.aiRecommendations ?? null,
    error_message: audit.errorMessage ?? null,
    error_code: audit.errorCode ?? null,
    started_at: audit.startedAt,
    completed_at: audit.completedAt ?? null,
    previous_score: previousScore ?? null,
    score_change: scoreChange,
  };
}

async function getPreviousScore(audit: AuditDoc): Promise<number | null> {
  if (!audit.websiteId) return null;
  const prev = await collections()
    .audits.find({
      websiteId: audit.websiteId,
      _id: { $ne: audit._id },
      status: "completed",
      overallScore: { $ne: null },
    })
    .sort({ completedAt: -1 })
    .limit(1)
    .next();
  return prev?.overallScore ?? null;
}

async function getAuditForAccess(publicId: string, userId: string | null | undefined) {
  const audit = await collections().audits.findOne({ publicId });
  if (!audit) return { error: { status: 404, message: "Audit not found." } };
  if (audit.userId && (!userId || audit.userId.toString() !== userId)) {
    return { error: { status: 403, message: "You do not have access to this audit." } };
  }
  return { audit };
}

auditsRouter.use(optionalAuth);

auditsRouter.post(
  "/",
  asyncHandler(async (req: AuthRequest, res) => {
    const key = clientKey(req);
    if (!auditPerMinute.allow(key)) {
      res.status(429).json({ detail: "You are starting audits too quickly. Please wait a moment." });
      return;
    }
    if (!auditPerHour.allow(key)) {
      res.status(429).json({ detail: "You have reached the hourly audit limit. Please try again later." });
      return;
    }

    const { url: rawUrl, language } = req.body || {};
    const lang = String(language || "en").trim().toLowerCase();
    if (!["en", "uz", "ru"].includes(lang)) {
      res.status(422).json({ detail: "Unsupported language." });
      return;
    }
    let validatedUrl: string;
    try {
      const validated = await validateUrl(String(rawUrl || ""));
      validatedUrl = validated.url;
    } catch (err) {
      if (err instanceof URLValidationError) {
        res.status(422).json({ detail: err.safeMessage });
        return;
      }
      throw err;
    }

    let maxPages = config.maxPages;
    if (req.user) {
      const startOfMonth = new Date();
      startOfMonth.setUTCDate(1);
      startOfMonth.setUTCHours(0, 0, 0, 0);
      const used = await collections().audits.countDocuments({
        userId: req.user._id,
        startedAt: { $gte: startOfMonth },
      });
      if (used >= config.monthlyAuditLimit) {
        res.status(403).json({
          detail: {
            code: "monthly_limit_reached",
            message: "You have reached your monthly audit limit.",
          },
        });
        return;
      }
    }

    const audit: AuditDoc = {
      _id: new ObjectId(),
      publicId: randomUUID(),
      userId: req.user?._id ?? null,
      websiteId: null,
      url: validatedUrl,
      status: "queued",
      progress: 0,
      stage: "Queued",
      maxPages,
      language: lang,
      overallScore: null,
      categoryScores: null,
      summary: null,
      results: null,
      aiRecommendations: null,
      errorMessage: null,
      errorCode: null,
      startedAt: new Date(),
      completedAt: null,
    };
    await collections().audits.insertOne(audit);

    if (!enqueueAudit(audit._id.toString())) {
      await collections().audits.updateOne(
        { _id: audit._id },
        {
          $set: {
            status: "failed",
            errorMessage: "Our auditors are busy right now. Please try again in a moment.",
            errorCode: "crawl_failed",
          },
        }
      );
      audit.status = "failed";
      audit.errorMessage = "Our auditors are busy right now. Please try again in a moment.";
      audit.errorCode = "crawl_failed";
    }

    res.status(201).json(serializeAudit(audit));
  })
);

auditsRouter.get(
  "/",
  requiredAuth,
  asyncHandler(async (req: AuthRequest, res) => {
    const audits = await collections()
      .audits.find({ userId: req.user!._id })
      .sort({ startedAt: -1 })
      .limit(100)
      .toArray();
    const items = [];
    for (const audit of audits) {
      const previous = await getPreviousScore(audit);
      items.push({
        public_id: audit.publicId,
        url: audit.url,
        status: audit.status,
        overall_score: audit.overallScore ?? null,
        started_at: audit.startedAt,
        score_change:
          audit.overallScore != null && previous != null ? audit.overallScore - previous : null,
        partial: Boolean(audit.results && (audit.results as { partial?: boolean }).partial),
        error_code: audit.errorCode ?? null,
        error_message: audit.errorMessage ?? null,
      });
    }
    res.json(items);
  })
);

auditsRouter.get(
  "/stats",
  requiredAuth,
  asyncHandler(async (req: AuthRequest, res) => {
    const audits = await collections()
      .audits.find({ userId: req.user!._id })
      .toArray();
    const completed = audits.filter((a) => a.status === "completed" && a.overallScore != null);
    const scores = completed.map((a) => a.overallScore as number);
    const recent = [...audits].sort((a, b) => b.startedAt.getTime() - a.startedAt.getTime()).slice(0, 5);
    res.json({
      total_audits: audits.length,
      completed_audits: completed.length,
      average_score: scores.length ? Math.round(scores.reduce((s, n) => s + n, 0) / scores.length) : null,
      best_score: scores.length ? Math.max(...scores) : null,
      recent_audits: recent.map((a) => ({
        public_id: a.publicId,
        url: a.url,
        status: a.status,
        overall_score: a.overallScore ?? null,
        started_at: a.startedAt,
        partial: Boolean(a.results && (a.results as { partial?: boolean }).partial),
        error_code: a.errorCode ?? null,
      })),
    });
  })
);

auditsRouter.get(
  "/:publicId",
  asyncHandler(async (req: AuthRequest, res) => {
    const { audit, error } = await getAuditForAccess(req.params.publicId, req.userId);
    if (error || !audit) {
      res.status(error?.status || 404).json({ detail: error?.message || "Audit not found." });
      return;
    }
    res.json(serializeAudit(audit, await getPreviousScore(audit)));
  })
);

auditsRouter.get(
  "/:publicId/report",
  asyncHandler(async (req: AuthRequest, res) => {
    const { audit, error } = await getAuditForAccess(req.params.publicId, req.userId);
    if (error || !audit) {
      res.status(error?.status || 404).json({ detail: error?.message || "Audit not found." });
      return;
    }
    if (audit.status !== "completed") {
      res.status(409).json({ detail: "This audit has not finished yet." });
      return;
    }
    res.json(serializeAudit(audit, await getPreviousScore(audit)));
  })
);

auditsRouter.delete(
  "/:publicId",
  requiredAuth,
  asyncHandler(async (req: AuthRequest, res) => {
    const { audit, error } = await getAuditForAccess(req.params.publicId, req.userId);
    if (error || !audit) {
      res.status(error?.status || 404).json({ detail: error?.message || "Audit not found." });
      return;
    }
    if (!audit.userId || audit.userId.toString() !== req.userId) {
      res.status(403).json({ detail: "You can only delete your own audits." });
      return;
    }
    await collections().audits.deleteOne({ _id: audit._id });
    res.json({ ok: true });
  })
);
